### PR TITLE
Allow retrying when an `EndpointGroup` is empty.

### PR DESCRIPTION
**Pull request by trustin**
Related: #1910

Motivation:

A client request to an empty `EndpointGroup` fails immediately with an
`EmptyEndpointGroupException`, even if there is a `Retrying(Rpc)Client`
in the decorator chain. As a result, a user cannot recover from the case
of an empty `EndpointGroup`.

Modifications:

- Change the contract of `EndpointSelector.select()` to return `null`
  when there's no `Endpoint` to select instead of throwing an
  `EmptyEndpointGroupException`.
- Make `DefaultClientRequestContext` handle the case where 
  `EndpointGroup.select()` returns `null` so that `init()` does not treat
  it as failure.
- Make `AbstractRetryingClient` handle the case where
  `EndpointGroup.select()` returns `null` so that it can continue retrying.
- Modify the signature of `ClientFactory.acquireEventLoop()` and
  `EventLoopScheduler.acquire()` so that it's possible to acquire an
  `EventLoop` even when `Endpoint` is not determined yet.

Result:

- A user can retry even when the current `EndpointGroup` is empty.
- (Breaking) `EndpointSelector.select()` (and thus `EndpointGroup.select()`
  as well) now returns `null` instead of throwing an `EmptyEndpointGroupException`.
- (Breaking) The method signature of `ClientFactory.acquireEventLoop()` and
  `EventLoopScheduler.acquire()` has been changed.

### CODE REVIEW

File: brave/src/main/java/com/linecorp/armeria/server/brave/ServiceRequestContextAdapter.java [hunk 1]
```diff
@@ -70,13 +68,6 @@ public String method() {
             return ctx.method().name();
         }
 
-        /**
-         * Original implementation is calling {@link HttpServerAdapter#url(Object)} which needs {@link
-         * RequestLog#scheme()}, but because {@link RequestLog#scheme()} is not always available, we need to
-         * use {@link RequestContext#path()} directly.
-         *
-         * @see brave.http.HttpServerRequest#path()
-         */
```
Comment (trustin) [inline brave/src/main/java/com/linecorp/armeria/server/brave/ServiceRequestContextAdapter.java:79]: Unrelated with this PR, but this removes a deprecation warning during compilation. The comment block seems not relevant anymore.

File: brave/src/main/java/com/linecorp/armeria/server/brave/ServiceRequestContextAdapter.java [hunk 2]
```diff
@@ -56,8 +55,7 @@
          */
         @Override
         public boolean parseClientIpAndPort(Span span) {
-            return parseClientIpFromXForwardedFor(span) ||
-                SpanTags.updateRemoteEndpoint(span, ctx);
```
Comment (trustin) [inline brave/src/main/java/com/linecorp/armeria/server/brave/ServiceRequestContextAdapter.java:60]: Unrelated with this PR - just fixed indentation

File: brave/src/main/java/com/linecorp/armeria/server/brave/ServiceRequestContextAdapter.java [hunk 3]
```diff
@@ -118,8 +109,8 @@ public long startTimestamp() {
         }
     }
 
-    static brave.http.HttpServerResponse asHttpServerResponse(RequestLog log,
-        brave.http.HttpServerRequest request) {
```
Comment (trustin) [inline brave/src/main/java/com/linecorp/armeria/server/brave/ServiceRequestContextAdapter.java:122]: Unrelated with this PR - just fixed indentation

**Comment by trustin** [comment]
> Should we make `EndpointGroup.select()` return `null` instead of throwing an `EmptyEndpointGroupException`? i.e. Instantiate `EmptyEndpointGroupException` only when setting `responseCause`.

If there's an objection, I'd like to make `EndpointGroup.select()` to return `null` instead of raising an exception.

File: core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java [hunk 4]
```diff
@@ -62,12 +58,22 @@
...
+            // - `EndpointGroup.select()` raised an `EmptyEndpointGroupException`.
+            // - Other exception was raised while context initialization.
+            //
+            // Because all the clean-up is done by `DefaultClientRequestContext.failEarly()`
+            // when context initialization fails with an exception other than `EmptyEndpointGroupException`,
+            // we can assume that the exception and response created here will be exposed only
+            // when context initialization failed due to an `EmptyEndpointGroupException`.
+            //
+            // See `DefaultClientRequestContext.init()` for more information.
+            System.err.println("!!!!!!");
```
Comment (andrey-tpt) [inline core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java:72]: Is the write to stderr here on purpose or a leftover after a debugging session?

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java]: !!!!!!

**Comment by ikhoon** [comment]
> Should we make EndpointGroup.select() return null instead of throwing an EmptyEndpointGroupException? i.e. Instantiate EmptyEndpointGroupException only when setting responseCause.

+1. If a result is empty, `null` sounds good to me.

File: core/src/main/java/com/linecorp/armeria/client/ClientFactory.java [hunk 5]
```diff
@@ -161,8 +162,15 @@ static void disableShutdownHook() {
...
+     *
+     * @param sessionProtocol the {@link SessionProtocol} of the connection
+     * @param endpointGroup the {@link EndpointGroup} where {@code endpoint} belongs to.
+     * @param endpoint the {@link Endpoint} where a request is being sent.
+     *                 {@code null} if the {@link Endpoint} is not known yet.
      */
-    ReleasableHolder<EventLoop> acquireEventLoop(Endpoint endpoint, SessionProtocol sessionProtocol);
+    ReleasableHolder<EventLoop> acquireEventLoop(SessionProtocol sessionProtocol,
+                                                 EndpointGroup endpointGroup,
+                                                 @Nullable Endpoint endpoint);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactory.java:173]: Maybe, Is this a breaking change? Need a `breaking-changes` label?

File: core/src/main/java/com/linecorp/armeria/client/ClientRequestContext.java [hunk 6]
```diff
@@ -252,7 +250,7 @@ default ClientRequestContext newDerivedContext(RequestId id,
      * <p>Note that this method does not copy the {@link RequestLog} properties to the derived context.
      */
     ClientRequestContext newDerivedContext(RequestId id, @Nullable HttpRequest req, @Nullable RpcRequest rpcReq,
-                                           Endpoint endpoint);
+                                           @Nullable Endpoint endpoint);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/ClientRequestContext.java:253]: Ditto.

File: core/src/main/java/com/linecorp/armeria/client/ClientFactory.java (hunk 5 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/ClientFactory.java:173]: Good idea. Added the label.

**Pull request merged**